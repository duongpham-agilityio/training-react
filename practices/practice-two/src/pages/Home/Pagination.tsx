import { memo } from 'react';
import isEqual from 'react-fast-compare';

// Components
import { Button } from 'components/commons';

// Styles
import homeStyles from 'pages/Home/index.module.css';

export interface PaginationProps {
  currentPage: number;
  pagination: number[];
  onChangePage: (page: number) => void;
}

const Pagination = ({
  pagination,
  currentPage,
  onChangePage,
}: PaginationProps) => {
  return (
    <div className={homeStyles.pagination}>
      {pagination.map((_, index) => {
        const page = index + 1;
        const handleChangePage = () => {
          onChangePage(page);
        };

        return (
          <Button
            label={`${page}`}
            variant={currentPage === page ? 'primary' : 'secondary'}
            size="small"
            key={index}
            onClick={handleChangePage}
            data-testid="pagination"
          />
        );
      })}
    </div>
  );
};

const areCompare = (prev: PaginationProps, next: PaginationProps): boolean =>
  isEqual(prev.currentPage, next.currentPage) &&
  isEqual(prev.pagination, next.pagination);

export default memo(Pagination, areCompare);
