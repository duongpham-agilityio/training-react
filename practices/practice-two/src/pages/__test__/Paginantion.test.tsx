// Mocks
import 'components/__test__/mocks/image.test';

import { fireEvent, render } from '@testing-library/react';

// COmponents
import Pagination, { PaginationProps } from 'pages/Home/Pagination';

const Component = (props: PaginationProps) => <Pagination {...props} />;

const setup = (props: PaginationProps) => render(<Component {...props} />);

jest.mock('react-fast-compare', () => ({
  default: jest.fn(),
}));

describe('Pagination', () => {
  it('Match to snapshot', () => {
    const { container } = setup({
      currentPage: 1,
      pagination: [0, 1, 2],
      onChangePage: jest.fn(),
    });

    expect(container).toMatchSnapshot();
  });

  it('Change pagination', () => {
    const changePage = jest.fn();
    const { rerender, getAllByRole } = setup({
      currentPage: 1,
      pagination: [0, 1, 2],
      onChangePage: changePage,
    });

    rerender(
      <Component
        {...{
          currentPage: 1,
          pagination: [0, 1, 2, 3],
          onChangePage: changePage,
        }}
      />
    );
    let buttons = getAllByRole('button');

    expect(buttons.length).toBe(4);

    rerender(
      <Component
        {...{
          currentPage: 1,
          pagination: [0, 1, 2, 3],
          onChangePage: changePage,
        }}
      />
    );

    buttons = getAllByRole('button');

    expect(buttons.length).toBe(4);

    fireEvent.click(buttons[1]);

    expect(changePage).toBeCalled();
  });
});
