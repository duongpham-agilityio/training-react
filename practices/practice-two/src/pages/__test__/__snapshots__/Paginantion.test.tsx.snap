// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination Match to snapshot 1`] = `
<div>
  <div>
    <button
      class="undefined undefined
  function small() { [native code] } undefined
  undefined "
      data-testid="pagination"
      type="button"
    >
      1
    </button>
    <button
      class="undefined undefined
  function small() { [native code] } undefined
  undefined "
      data-testid="pagination"
      type="button"
    >
      2
    </button>
    <button
      class="undefined undefined
  function small() { [native code] } undefined
  undefined "
      data-testid="pagination"
      type="button"
    >
      3
    </button>
  </div>
</div>
`;
